### ISSUE TITLE
Doing a re-register with same socket address/ same endpoint will break registration by address index

### ISSUE THREAD

**Issue by acaibird** [eclipse/leshan#625]
```
    /* *************** Leshan Registration API **************** */
    ```
@Override
    public Deregistration addRegistration(Registration registration) {
        try {
            lock.writeLock().lock();
            Registration registrationRemoved = regsByEp.put(registration.getEndpoint(), registration);
           regsByRegId.put(registration.getId(), registration);
           // If a registration is already associated to this address we don't care as we only 
           //want to keep the most recent binding.
            regsByAddr.put(registration.getSocketAddress(), registration);
            if (registrationRemoved != null) {
                Collection<Observation> observationsRemoved = 
                                                                   unsafeRemoveAllObservations(registrationRemoved.getId());
                removeFromMap(regsByAddr, registrationRemoved.getSocketAddress(), registrationRemoved);
                removeFromMap(regsByRegId, registrationRemoved.getId(), registrationRemoved);
                return new Deregistration(registrationRemoved, observationsRemoved);
            }
        } finally {
            lock.writeLock().unlock();
        }
        return null;
    }
```
In this registration process, why add registration information before delete it? When I try to **keep the client's registrationId unchanged**, the client registers twice, and the client's information will be deleted rather than refreshed. 
So the following exception is thrown:
java.lang.IllegalStateException: no registration for this Id.......

**Comment by acaibird** [comment]
```
@Override
	public String getRegistrationId(RegisterRequest registerRequest) {
		// TODO Auto-generated method stub
		return registerRequest.getEndpointName();
	}
```
I use device's endpoint name as registrationId rather than randomly generated. 
Why do I register the device twice? Considering the automatic registration when the device is restarted on boot， if the device has been registered, but the device is restarted.

**Comment by sbernard31** [comment]
> I use device's endpoint name as registrationId rather than randomly generated.

This is out of the specification as you will not be able to make difference between 2 different registrations for the same endpoint ...

>  Considering the automatic registration when the device is restarted on boot， if the device has been registered, but the device is restarted.

In that case this is a new registration. (Register request means new registration) if your device want to keep the same registration, it must send an UPDATE request not a REGISTER one. 

This is how I understand the specification. Why do you want to use endpoint name as registration Id ? and why is it a problem to register ?

**Comment by acaibird** [comment]
But what if the two registrations of the device are the same **IP:PORT**? 
The first registration of the device is successful, but for some reason the device is restarted. The device sends a new registration to the server. Unfortunately, the device's IP and port remain unchanged. In this way, the device receives feedback on successful registration. But apparently on the server side, the Map  **regsByAddr** has no information about it. 
Am I right in this analysis? Thanks.

**Comment by sbernard31** [comment]
> what if the two registrations of the device are the same IP:PORT

The old registration MUST be deleted and a new one recreate with this IP:PORT

> But apparently on the server side, the Map regsByAddr has no information about it.

I will look at this deeper, but I suspect that you maybe found a bug.

**Comment by sbernard31** [comment]
I confirm you found a bug. I tried to fix it in master (4ad71b1). I inspired myself from [Redis implemention of registration store](https://github.com/eclipse/leshan/blob/master/leshan-server-cluster/src/main/java/org/eclipse/leshan/server/cluster/RedisRegistrationStore.java#L160).

Thx for reporting this ! If you could try the fix and tell us if it works for you now, this will be really appreciated.

(Anyway, keep in mind that using same regId for 2 different registrations is out of spec and this is probably not a good idea)

**Comment by acaibird** [comment]
```
 /* *************** Leshan Registration API **************** */
    @Override
    public Deregistration addRegistration(Registration registration) {
    	 Registration registrationRemoved = null;
    	 Collection<Observation> observationsRemoved = null;
        try {
            lock.writeLock().lock();
            registrationRemoved = regsByEp.put(registration.getEndpoint(), registration);
            // If a registration is already associated to this address we don't care as we only 
            //want to keep the mostrecent binding.
            if (registrationRemoved != null) {
                observationsRemoved = unsafeRemoveAllObservations(registrationRemoved.getId());
                removeFromMap(regsByAddr, registrationRemoved.getSocketAddress(), registrationRemoved);
                removeFromMap(regsByRegId, registrationRemoved.getId(), registrationRemoved);
            }
            regsByRegId.put(registration.getId(), registration);
            regsByAddr.put(registration.getSocketAddress(), registration); 
            if(registrationRemoved!=null) {
            	return new Deregistration(registrationRemoved, observationsRemoved);
            }     
        } finally {
            lock.writeLock().unlock();
        }
        return null;
    }
```
All I did was adjust the order of additions and deletions. I delete the old Registration Info first, and then add new one.

**Comment by sbernard31** [comment]
I would prefer to avoid to delete then re-add again and I would like to have the redis implementation and the In memory one as close as possible.

I plan to release a new version of Leshan today.

**Comment by sbernard31** [comment]
(it should be fixed in master commit 4ad71b1)
